File: agent/src/fa-align.mjs
#!/usr/bin/env node
/**
 * フォースドアライメント（Demucs分離＋torchaudio MMS_FA）で単語ごとの秒数を出す。
 * whisper文字起こしと違い「既知の歌詞」を音声に強制整列するため、rap・反復コーラスでも
 * 取り違えが起きにくく、語間の「間」も正確に取れる（＝gap分割の土台）。
 *
 * 出力: agent/{slug}/assets/fa_words.json  … cueごとの [{w,s,e}]
 * 依存: agent/.fa-venv（無ければ自動作成し torch/torchaudio/demucs/soundfile を入れる）
 *
 * Usage:
 *   node agent/src/fa-align.mjs --slug <slug>                 # full-cues.json  → fa_words.json
 *   node agent/src/fa-align.mjs --slug <slug> --source lines  # full-lines.json → fa_words_lines.json
 *
 * --source lines は「キューを作る前の生の歌詞行」に整列する。意味分割(semantic-chunk.mjs)は
 * この行単位の語秒を土台にするので、既存キューの切れ目に縛られない。
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { spawnSync } from "child_process";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const AGENT = path.resolve(__dirname, "..");
const REPO = path.resolve(AGENT, "..");
const VENV = path.join(AGENT, ".fa-venv");
const PY = path.join(VENV, "bin", "python");
const PY_SYS = "/opt/homebrew/bin/python3.12";

const getArg = (n, d = null) => { const i = process.argv.indexOf(`--${n}`); return i >= 0 && process.argv[i + 1] ? process.argv[i + 1] : d; };
const slug = getArg("slug");
if (!slug) { console.error("Usage: node agent/src/fa-align.mjs --slug <slug>"); process.exit(1); }

const source = getArg("source", "cues");
const SRC = source === "lines" ? "full-lines.json" : "full-cues.json";
const OUT = source === "lines" ? "fa_words_lines.json" : "fa_words.json";

const assets = path.join(AGENT, slug, "assets");
if (!fs.existsSync(path.join(assets, SRC))) { console.error(`[fa-align] ${assets}/${SRC} が無い`); process.exit(2); }

function run(cmd, cmdArgs, opts = {}) {
  const r = spawnSync(cmd, cmdArgs, { stdio: "inherit", ...opts });
  if (r.status !== 0) { console.error(`[fa-align] 失敗: ${cmd} ${cmdArgs.join(" ")}`); process.exit(r.status || 1); }
}

if (!fs.existsSync(PY)) {
  if (!fs.existsSync(PY_SYS)) { console.error(`[fa-align] ${PY_SYS} が無い（brew install python@3.12）`); process.exit(3); }
  console.log("[fa-align] .fa-venv を作成中（初回のみ・torch等DL）...");
  run(PY_SYS, ["-m", "venv", VENV]);
  run(PY, ["-m", "pip", "install", "-q", "--upgrade", "pip"]);
  run(PY, ["-m", "pip", "install", "torch", "torchaudio", "demucs", "soundfile"]);
}

run(PY, [path.join(AGENT, "src", "fa-align.py"), assets, SRC, OUT], { cwd: REPO });
console.log(`[fa-align] OK -> agent/${slug}/assets/${OUT}`);

File: agent/src/fa-align.py
#!/usr/bin/env python3
# Forced alignment for full-lyric videos.
# Demucs vocal separation -> torchaudio MMS_FA forced alignment against known lyrics.
# Outputs per-word start/end seconds so the cue-editor can split at real pauses
# and place fragment start times precisely. Does NOT transcribe (lyrics are known).
#
# Usage: fa-align.py <assetsDir> [srcJson=full-cues.json] [outJson=fa_words.json]
#   reads   <assetsDir>/<srcJson>       (array of {eng,...} — cues でも lines でも可)
#   reads   <assetsDir>/audio-full.mp3  (or audio.mp3)
#   writes  <assetsDir>/<outJson>       (array-per-entry of {w,s,e})
#   caches  <assetsDir>/.stems/vocals.wav
import json, re, sys, os, subprocess
import torch, torchaudio
import soundfile as sf

assets = sys.argv[1]
src_json = sys.argv[2] if len(sys.argv) > 2 else "full-cues.json"
out_json = sys.argv[3] if len(sys.argv) > 3 else "fa_words.json"
cues = json.load(open(os.path.join(assets, src_json)))
audio = next((os.path.join(assets, n) for n in ("audio-full.mp3", "audio.mp3", "audio-full.wav")
             if os.path.exists(os.path.join(assets, n))), None)
if not audio:
    print("NO_AUDIO", file=sys.stderr); sys.exit(2)

# --- vocal separation (cached) ---
stem_dir = os.path.join(assets, ".stems")
base = os.path.splitext(os.path.basename(audio))[0]
vocals = os.path.join(stem_dir, "htdemucs", base, "vocals.wav")
if not os.path.exists(vocals):
    print("[demucs] separating vocals...", flush=True)
    subprocess.run([sys.executable, "-m", "demucs", "--two-stems=vocals",
                    "-o", stem_dir, audio], check=True)
src = vocals if os.path.exists(vocals) else audio

data, sr = sf.read(src, dtype="float32", always_2d=True)
wav = torch.from_numpy(data.T).mean(0, keepdim=True)
if sr != 16000:
    wav = torchaudio.functional.resample(wav, sr, 16000); sr = 16000

from torchaudio.pipelines import MMS_FA as bundle
model = bundle.get_model()
tokenizer = bundle.get_tokenizer()
aligner = bundle.get_aligner()
DICT = bundle.get_dict()

def norm(s):
    s = (s or "").lower().replace("’", "'")
    s = re.sub(r"[^a-z' ]", " ", s)
    return [w for w in s.split() if any(c in DICT for c in w)]

transcript, counts = [], []
for c in cues:
    ws = norm(c.get("eng", "")); counts.append(len(ws)); transcript += ws

print(f"[align] {len(transcript)} words on {os.path.basename(src)}", flush=True)
with torch.inference_mode():
    emission, _ = model(wav)
    spans = aligner(emission[0], tokenizer(transcript))
ratio = wav.shape[1] / emission.shape[1] / sr

se = [(s[0].start * ratio, s[-1].end * ratio) for s in spans]
out, idx = [], 0
for i, c in enumerate(cues):
    n = counts[i]; wl = norm(c.get("eng", ""))
    out.append([{"w": wl[k], "s": round(se[idx+k][0], 3), "e": round(se[idx+k][1], 3)} for k in range(n)])
    idx += n

json.dump(out, open(os.path.join(assets, out_json), "w"))
print(f"[done] wrote {out_json} ({len(out)} entries)")
